### ISSUE TITLE
how can I run network in python?

### ISSUE THREAD

**Issue by justinbuzzni** [suyogduttjain/pixelobjectness#1]
I tried to run pixelobjectness network in python.
following code is the code.

The output of following code is not the expected result.

I want to know what is the wrong code.

`
import numpy as np
from PIL import Image as PILImage
import caffe
_FOLDER = '/opt/pixelobjectness/'
MODEL_FILE = _FOLDER+'test.prototxt'
PRETRAINED = _FOLDER+'pixel_objectness.caffemodel'
net = caffe.Net(MODEL_FILE, PRETRAINED)

net.set_mode_gpu()
net.set_device(0)
net.set_phase_test()

mean_vec = np.array([104.008, 116.669, 122.675], dtype=np.float32)
reshaped_mean_vec = mean_vec.reshape(1, 1, 3);

net.set_raw_scale('data',255)
net.set_mean('data', reshaped_mean_vec)
net.set_channel_swap('data', (2, 1, 0))

IMAGE_FILE = '/opt/pixelobjectness/images/n01736375_3996.JPEG'
im = caffe.io.load_image(IMAGE_FILE)
im2 = net.preprocess('data',im)

net.blobs['data'].data[...] = im2

output = net.forward()
d = output['fc8_interp'].reshape(513,513,2,1)

import cPickle as pickle
with file("/root/workspace/images/a2.mat","w") as fout:
    pickle.dump(output['fc8_interp'], fout)

`

**Comment by justinbuzzni** [comment]
I resolved the problem using the following code.
`
# -*- coding: utf-8 -*-
"""
This package contains code for the "CRF-RNN" semantic image segmentation method, published in the
ICCV 2015 paper Conditional Random Fields as Recurrent Neural Networks. Our software is built on
top of the Caffe deep learning library.

Contact:
Shuai Zheng (szheng@robots.ox.ac.uk), Sadeep Jayasumana (sadeep@robots.ox.ac.uk), Bernardino Romera-Paredes (bernard@robots.ox.ac.uk)

Supervisor:
Philip Torr (philip.torr@eng.ox.ac.uk)

For more information about CRF-RNN, please vist the project website http://crfasrnn.torr.vision.
"""

import sys, getopt

import os
import cPickle
import logging
import numpy as np
import pandas as pd
from PIL import Image as PILImage
#import Image
import cStringIO as StringIO
sys.path.append('/opt/caffe/python/')

import caffe
import matplotlib.pyplot as plt
_FOLDER = '/opt/pixelobjectness/'
MODEL_FILE = _FOLDER+'test.prototxt'
PRETRAINED = _FOLDER+'pixel_objectness.caffemodel'
net = caffe.Net(MODEL_FILE, PRETRAINED)
net.set_mode_gpu()
net.set_device(0)



def tic():
    #http://stackoverflow.com/questions/5849800/tic-toc-functions-analog-in-python
    #Homemade version of matlab tic and toc functions
    import time
    global startTime_for_tictoc
    startTime_for_tictoc = time.time()

def toc():
    import time
    if 'startTime_for_tictoc' in globals():
        print "Elapsed time is " + str(time.time() - startTime_for_tictoc) + " seconds."
    else:
        print "Toc: start time not set"


def getpallete(num_cls):
        # this function is to get the colormap for visualizing the segmentation mask
        n = num_cls
        pallete = [0]*(n*3)
        for j in xrange(0,n):
                lab = j
                pallete[j*3+0] = 0
                pallete[j*3+1] = 0
                pallete[j*3+2] = 0
                i = 0
                while (lab > 0):
                        pallete[j*3+0] |= (((lab >> 0) & 1) << (7-i))
                        pallete[j*3+1] |= (((lab >> 1) & 1) << (7-i))
                        pallete[j*3+2] |= (((lab >> 2) & 1) << (7-i))
                        i = i + 1
                        lab >>= 3
        return pallete


def crfasrnn_segmenter(model_file, pretrained_file, gpudevice, inputs):

    input_ = np.zeros((len(inputs),
        500, 500, inputs[0].shape[2]),
        dtype=np.float32)
    for ix, in_ in enumerate(inputs):
        input_[ix] = in_

    # Segment
    caffe_in = np.zeros(np.array(input_.shape)[[0,3,1,2]],
                        dtype=np.float32)
    for ix, in_ in enumerate(input_):
        caffe_in[ix] = in_.transpose((2, 0, 1))
    tic()
    out = net.forward_all(**{net.inputs[0]: caffe_in})
    toc()
    predictions = out[net.outputs[0]]

    return predictions[0].argmax(axis=0).astype(np.uint8)

def run_crfasrnn(inputfile, outputfile, gpudevice):

    IMAGE_FILE = inputfile

    input_image = 255 * caffe.io.load_image(IMAGE_FILE)
    input_image = resizeImage(input_image)

    width = input_image.shape[0]
    height = input_image.shape[1]
    maxDim = max(width,height)

    image = PILImage.fromarray(np.uint8(input_image))
    image = np.array(image)

    pallete = getpallete(256)

    mean_vec = np.array([103.939, 116.779, 123.68], dtype=np.float32)
    reshaped_mean_vec = mean_vec.reshape(1, 1, 3);

    # Rearrange channels to form BGR
    im = image[:,:,::-1]
    # Subtract mean
    im = im - reshaped_mean_vec

    # Pad as necessary
    cur_h, cur_w, cur_c = im.shape
    pad_h = 500 - cur_h
    pad_w = 500 - cur_w
    im = np.pad(im, pad_width=((0, pad_h), (0, pad_w), (0, 0)), mode = 'constant', constant_values = 0)
    # Get predictions
    #segmentation = net.predict([im])
    segmentation  = crfasrnn_segmenter(MODEL_FILE,PRETRAINED,gpudevice,[im])

    segmentation2 = segmentation[0:cur_h, 0:cur_w]

    output_im = PILImage.fromarray(segmentation2)
    output_im.putpalette(pallete)
    output_im.convert('RGB').save(outputfile)
    return segmentation2

def resizeImage(image):
        width = image.shape[0]
        height = image.shape[1]
        maxDim = max(width,height)
        if maxDim>500:
            if height>width:
                ratio = float(500.0/height)
            else:
                ratio = float(500.0/width)
            image = PILImage.fromarray(np.uint8(image))
            image = image.resize((int(height*ratio), int(width*ratio)),resample=PILImage.BILINEAR)
            image = np.array(image)
        return image

import json


from PIL import Image
from StringIO import StringIO
def imgopen(url, timeout=10):
    """ read from an url without creating any file.
    """
    try:
        response = requests.get(url, timeout=timeout)
        img = Image.open(StringIO(response.content))
    except Exception, e:
        print e
        img = None
    return img

def _fit_size(w, h, s):
    """ fit width and height to size
        w: width
        h: height
        s: size
    """
    if h < w:
        w = int((float(w) / h) * s)
        h = s
    else:
        h = int((float(h) / w) * s)
        w = s
    return w, h

def thumbnail(url, fname, size=0, quality=100):
    """ create a thumbnail with the given size.
    """
    h = w = 0
    img = imgopen(url)
    w, h = img.size
    if size:
        w, h = _fit_size(w, h, size)
    try:
        img = img.convert("RGB") #png 형태로 된 이미지를 jpg로 구울때 필요함. 엄청 중요함.
        img.thumbnail((w, h), Image.ANTIALIAS) #ANTIALIAS option make image high quality and low size. it's best.
        img.save(fname, quality=quality)
        w, h = img.size
    except Exception, e:
        print "[ERROR]\t", e
        print "\t", url
        w = h = 0

    return w, h

if __name__ == '__main__':
    outputfile = "/tmp/output.jpg"
    inputfile =  '/opt/pixelobjectness/images/n02129530_1632.JPEG'

    segmentation = run_crfasrnn(inputfile, outputfile, 0)
    # httpd = simple_server.make_server('0.0.0.0', 8901, app)
    # httpd.serve_forever()


`

**Comment by justinbuzzni** [comment]
I resolved that problem using crfasrnn code.
